[ML] Add DeBERTa-V2/V3 tokenizer

maxhniebergall opened the pull request:
This PR adds support for the DeBERTa-V2/V3 tokenizer (DeBERTa-V2 and V3 use the same tokenizer). It also adds the `balanced` truncation method. 

This PR contains:

- the new classes and the new named writables
- the tokens in Deberta are slightly different but other than that the code is almost identical to xlmRoberta
- I added the balanced truncation option which will now be available with all models which take two inputs
- byte fall back (decomposeBytePieces) for unknown tokens

Commit https://github.com/elastic/elasticsearch/pull/111852/commits/54f99952c882452a70f067fa6c1763eb821f4f53#diff-2f9a331e12c3fb7602085cb04501a93f424bafa029864c2f798b83860e9679b7 includes end-to-end testing of tokenization with a real vocab file from Huggingface, but it was too large to include in CI.

Commit:
start to add deberta-v2 tokenizer classes

Commit:
continue to add basic tokenizer stuff

Commit:
Finish adding DeBERTa-2 tokenizer

Still need to review & test

maxhniebergall committed:
Update docs/changelog/111852.yaml

elasticsearchmachine commented:
Hi @maxhniebergall, I've created a changelog YAML for you.

maxhniebergall commented on x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/DebertaV3TestVocab.java:
It might not be worth including this whole vocabulary file as it is 2.9MB and its not really necessary.

Commit:
Complete test setup and linting

Commit:
Add serialization of deberta tokenization

Commit:
fix request buillder to match model

Commit:
debugging

Commit:
add balanced truncation

Commit:
remove full vocabulary and use tiny vocab for tests

maxhniebergall commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/DebertaTokenizationResult.java:
```diff
@@ -0,0 +1,138 @@
+                tokenIds.add(IntStream.of(sepTokenId));
+                tokenMap.add(IntStream.of(SPECIAL_TOKEN_POSITION));
+            }
+            seqPairOffset = withSpecialTokens ? tokenId1s.size() + 2 : tokenId1s.size();
```
the final special token isn't included in the offset

Commit:
Remove TODO

elasticsearchmachine commented:
Pinging @elastic/ml-core (Team:ML)

Commit:
precommit

Commit:
Add named writables and known tokenizers

Commit:
Add deberta to list of known tokenizers in test

davidkyle commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/NlpTokenizer.java:
```diff
@@ -232,6 +234,19 @@ public TokenizationResult.Tokens tokenize(
                     tokenIdsSeq2 = tokenIdsSeq2.subList(0, maxSequenceLength() - extraTokens - tokenIdsSeq1.size());
                     tokenPositionMapSeq2 = tokenPositionMapSeq2.subList(0, maxSequenceLength() - extraTokens - tokenIdsSeq1.size());
                 }
+                case BALANCED -> {
```
Because `NlpTokenizer` is an abstract class most of the unit testing of this code is in `BertTokenizerTests`. Look at `BertTokenizerTests::testTokenizeLargeInputMultiSequenceTruncation` and add a similar test to cover the `BALANCED` case

maxhniebergall replied:
thanks!

davidkyle commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/DebertaV2Tokenizer.java:
```diff
@@ -0,0 +1,306 @@
+                this.scores[i++] = s;
+            }
+            normalizer = PrecompiledCharMapNormalizer.fromBase64EncodedResource(
+                "/org/elasticsearch/xpack/ml/inference.nlp.tokenizers/spm_precompiled_normalizer.txt"
```
This file is the sentence piece model for the XLM Roberta model added in #94089. Doesn't DeBERTa use a different model. If so we either have to include it as a resource or add the ability to install it

maxhniebergall replied:
I'll need to verify this in the end to end testing. I had tried to extract the normalizer from the spm file provided in huggingface but I wasn't able to do it. I'm also not sure that the normalizer is different.

maxhniebergall replied:
I haven't been able to complete the end-to-end testing yet because the cpp snapshot isn't available yet. I created a notebook to test it though.

maxhniebergall replied:
I just went through the steps of extracting the precompiled charsmap from the deberta SPM file and it is identical to this older one

maxhniebergall commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/NlpTokenizer.java:
```diff
@@ -232,6 +234,19 @@ public TokenizationResult.Tokens tokenize(
                 }
+                case BALANCED -> {
+                    isTruncated = true;
+                    if (tokenIdsSeq1.size() + tokenIdsSeq2.size() + getNumExtraTokensForSeqPair() > maxSequenceLength()) {
```
good point!

Commit:
Add tests for balanced tokenizer and fix errors in tokenizer logic

maxhniebergall commented:
@elasticmachine merge upstream

maxhniebergall commented:
theres an issue with the tokenizer to do with this special character `\xad` (which is a hidden "soft-hyphen").

In our system, that character is tokenized as `[1,336,2,507,3,2]`, but python is giving `[1, 336, 2, 507, 198, 177, 2]`. token 507 is the whitespace character, so I guess sentencepiece is interpreting the soft-hyphen as a whitespace character, added onto unknown character in ES and as `\xad` in python. In the vocabulary, 198 is <0xC2> and 177 is <0xAD>, which corresponds to the UTF-8 hex code for soft-hyphen. So in python, the character is being correctly broken down into the these two byte, but in our system it is being interpreted as one unknown character...

Commit:
fix order of parameters passed to deberta

Commit:
Add support for byte_fallback which is enabled for DeBERTa

byte_fallback decomposes unknown tokens into multiple tokens each of one byte if those bytes are in the vocabulary.

maxhniebergall commented:
I've tested the tokenizer on the first 6000 query-passage pairs in MS-MARCO and the differences in score between the direct-pytorch and elasticsearch versions are around 10^-6 at most (a few low 10^-5), and the total variance was `6.320295683227377e-12`, which is extremely low. I spot checked some of the cases and there wasn't any difference in the tokenization, so I believe this is due to errors in floating point arithmetic, as the orders of magnitude align with that. I'm going to run the test on the rest of MS-MARCO overnight to see if there are any other outliers.

Commit:
precommit

maxhniebergall commented:
@elasticmachine merge upstream

elasticmachine committed:
Merge branch 'main' into addDebertaTokenizer

Commit:
update tests to account for byte decomposition

Commit:
remove sysout

Commit:
fix tests for byteFallback, for real this time

maxhniebergall commented:
Manual testing revealed that any differences in tokenization are due to floating point rounding in scores and are not fixable.

davidkyle reviewed (approved):
LGTM

> Commit https://github.com/elastic/elasticsearch/commit/54f99952c882452a70f067fa6c1763eb821f4f53#diff-2f9a331e12c3fb7602085cb04501a93f424bafa029864c2f798b83860e9679b7 includes end-to-end testing of tokenization with a real vocab file from Huggingface, but it was too large to include in CI.

Thanks, this was really useful for verifying the tokenisation.

Commit:
Move defaultSpanForChunking into super class to avoid repitition

Commit:
simplify decomposeBytePieces

Commit:
Merge branch 'main' of github.com:elastic/elasticsearch into addDebertaTokenizer

elasticsearchmachine commented:
## 💔 Backport failed
The backport operation could not be completed due to the following error:
```
An unexpected error occurred when attempting to backport this PR.
```

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 111852`
